tests/performance_tests/performance_test.py

```diff
@@ -11,203 +12,463 @@
+
+    @dataclass
+    class PluginThreadOptResult:
+        recommendedThreadCount: int = 0
+        threadToMaxTpsDict: dict = field(default_factory=dict)
+        analysisStart: datetime = datetime.utcnow()
+        analysisFinish: datetime = datetime.utcnow()
+
+    def optimizePluginThreadCount(self,  optPlugin: PluginThreadOpt, optType: PluginThreadOptRunType=PluginThreadOptRunType.LOCAL_MAX,
+                                  minThreadCount: int=2, maxThreadCount: int=os.cpu_count()) -> PluginThreadOptResult:
+
+        if optPlugin == PerformanceTest.PluginThreadOpt.PRODUCER:
+            fileName = "producerThreadResults.txt"
+        elif optPlugin == PerformanceTest.PluginThreadOpt.CHAIN:
+            fileName = "chainThreadResults.txt"
         else:
-            removeArtifacts(testTimeStampDirPath)
-    except OSError as error:
-        print(error)
-
-def testDirsSetup(rootLogDir, testTimeStampDirPath, ptbLogsDirPath):
-    try:
-        def createArtifactsDir(path):
-            print(f"Checking if test artifacts dir exists: {path}")
-            if not os.path.isdir(f"{path}"):
-                print(f"Creating test artifacts dir: {path}")
-                os.mkdir(f"{path}")
-
-        createArtifactsDir(rootLogDir)
-        createArtifactsDir(testTimeStampDirPath)
-        createArtifactsDir(ptbLogsDirPath)
-
-    except OSError as error:
-        print(error)
-
-def prepArgsDict(testDurationSec, finalDurationSec, logsDir, maxTpsToTest, testIterationMinStep,
-                 tpsLimitPerGenerator, delReport, delTestReport, numAddlBlocksToPrune, quiet, delPerfLogs,
-                 testHelperConfig, testClusterConfig) -> dict:
-    argsDict = {}
-    argsDict.update(asdict(testHelperConfig))
-    argsDict.update(asdict(testClusterConfig))
-    argsDict.update({key:val for key, val in locals().items() if key in set(['testDurationSec', 'finalDurationSec', 'maxTpsToTest', 'testIterationMinStep', 'tpsLimitPerGenerator',
-                                                                             'delReport', 'delTestReport', 'numAddlBlocksToPrune', 'logsDir', 'quiet', 'delPerfLogs'])})
-    return argsDict
+            fileName = "netThreadResults.txt"
+
+        resultsFile = f"{self.loggingConfig.pluginThreadOptLogsDirPath}/{fileName}"
+
+        threadToMaxTpsDict: dict = {}
+
+        clusterConfig = copy.deepcopy(self.clusterConfig)
+        analysisStart = datetime.utcnow()
+
+        with open(resultsFile, 'w') as log:
+            log.write(f"{optPlugin.name.lower()}Threads, maxTpsAchieved\n")
+        log.close()
+
+        lastMaxTpsAchieved = 0
+        for threadCount in range(minThreadCount, maxThreadCount+1):
+            print(f"Running {optPlugin.name.lower()} thread count optimization check with {threadCount} {optPlugin.name.lower()} threads")
+
+            if optPlugin == PerformanceTest.PluginThreadOpt.PRODUCER:
+                clusterConfig.extraNodeosArgs.producerPluginArgs.producerThreads = threadCount
```

Comment:
If the thread count member of each PluginArgs class was named simply `threads` and the enum class values were appropriate strings rather than integers, this if/elif ladder could be a single line.  `getattr(clusterConfig, optPlugin.value + 'PluginArgs').threads = threadCount`.  Python introspection allows a sort of pseudo-object oriented style across otherwise unrelated classes that are also unaware of external enumerations and so don't have setters which use them.

Reply:
Good call. Thanks.
Addressed here: https://github.com/AntelopeIO/leap/pull/510/commits/0d5980d2da0e2af090e82710d4e592f5db820b8e